### DISCUSSION TITLE
Proposals for the design of the SQL database

### DISCUSSION THREAD

**Discussion by ACSG-64** [ACSG-64/concord-real-time-workspaces#4] [Development]
If you would like to suggest a database design, please share it here!

**ESSENTIAL APP SPECIFICATIONS:**

- People who wish to use the application must create their accounts and thus become users.
- Users can create workspaces and join other workspaces. The default channel of the workspace is `general`.
- Users (maybe only the owner) can create more channels in their workspace.
- Messages sent by users should persist.

_Comment if you require more details on the essential specifications or wish to make adjustments._

**Comment by ACSG-64** [comment]
Here is my proposal:
![Concord DB design diagram](https://user-images.githubusercontent.com/50815104/201506633-875b9071-d0af-4bfe-a91b-5b1e89457414.png)
Foreign keys in the diagram are implicit. Sequelize could generate them automatically when defining associations between models/entities.

**Comment by collengura** [comment]
That appears for fine to me. However, in message, I can see id and content. Does content include comments such as "reply"? If needed, where would we put that?

**Reply by collengura** [reply]
...appears fine...

**Comment by ACSG-64** [comment]
Diagram updated thanks to @collengura  and @tkshfj  suggestions.
- The relationship between message and reply has been added.
- Added timestamps as attributes for messages: `created_at `and `modified_at`, the latter in case the functionality to modify already sent messages is implemented.
![Concord DB design](https://user-images.githubusercontent.com/50815104/201696215-323b9f79-cabf-41e1-9561-5ca19c537506.png)

**Reply by tkshfj** [reply]
Looks good to me. Thank you.

**Comment by ACSG-64** [comment]
Relational diagram with foreign keys (ignore the datatypes)
![Concord DB relational diagram](https://user-images.githubusercontent.com/50815104/204144254-d7cb9610-3880-4998-8788-ed5d6bd43d6c.png)

**Comment by ACSG-64** [comment]
Small detail that I had forgotten. Workspaces should have an invitation id to invite other people via a link/URL (something like `worspaces/invitations/<invitation_id>`), and the owner should also have the ability to control whether that link is active or not to allow or prevent new people from joining the workspace
![Concord DB design](https://user-images.githubusercontent.com/50815104/204165497-e2f94bf1-6300-46f9-975a-c6e7955fdf2b.png)

**Comment by collengura** [comment]
Good morning Andre`s,This is getting better. Out of interest, what are you using to draw the diagram? Kind regards Collen Gura     From: Andrés SarmientoSent: Monday, 28 November 2022 01:30To: ACSG-64/concord-real-time-workspacesCc: Collen Gura; MentionSubject: Re: [ACSG-64/concord-real-time-workspaces] Proposals for the design of the SQL database (Discussion #4) Small detail that I had forgotten. Workspaces should have an invitation id to invite other people via a link/URL (something like worspaces/invitation/<invitation_id>), and the owner should also have the ability to control whether that link is active or not to allow or prevent new people from joining the workspace—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Reply by ACSG-64** [reply]
Hey @collengura I'm using [**Oracle SQL Developer Data Modeler**](https://www.oracle.com/database/sqldeveloper/technologies/sql-data-modeler/).

**Comment by ACSG-64** [comment]
Another update...

- The primary key for the 'membership' entity is now the composition between the user ID and the workspace ID to prevent duplicate data.
- The primary key for the 'channel' entity is now the composition between the workspace ID and the channel name, this to prevent the possibility of creating multiple channels with the same name in a workspace. The channel ID attribute will act as a secondary key and foreign key.

_The transferability of relationships is also explicitly drawn as a formality._

![Concord DB relational diagram v2](https://user-images.githubusercontent.com/50815104/205473164-637c5a2f-9c0a-4db5-996b-b399228c1c89.png)
